### PR TITLE
feat: add ManageSnapshots.fast_forward_branch

### PR DESCRIPTION
**Pull request by JeroenSchmidt**
# Rationale for this change
Adds `ManageSnapshots.fast_forward_branch(from_branch, to_ref)` to advance a
branch to the snapshot referenced by another branch or tag, provided the
current snapshot of `from_branch` is an ancestor of `to_ref`'s snapshot.

The method almost matches the behavior of Apache Iceberg's Java reference
implementation (`SnapshotManager.fastForwardBranch` /
`UpdateSnapshotReferencesOperation.replaceBranch(..., fastForward=true)`)
and, by extension, the Spark `system.fast_forward` stored procedure that
depends on it.

<strike>
However, while the implementation is very close there is one exception; 

Unlike Java's `ManageSnapshots.fastForwardBranch`, this method does not
provide Java-parity for intra-chain semantics.

1) Java maintains a mutable `updatedRefs` map that reflects prior operations
      in the same chain, so calling `createBranch` and then `fastForwardBranch`
      on the same ref in one chain observes the freshly-created state.

2) pyiceberg's `ManageSnapshots` accumulates `SetSnapshotRefUpdate` values without
      mutating `table_metadata.refs` between chained calls;
      -> The no-op and ancestry checks below operate on committed metadata only.
      -> Callers that need Java-parity behavior should commit between chain steps.

For the sake of scope I left out the changes needed to ensure chain sematic parity between the java and python.
Would the maintainers like a followup PR or have it included in this PR? My only worry is that it will increase the complexity of an otherwise straightforward PR. 
</strike>

Interchain semantics was introduced in https://github.com/apache/iceberg-python/pull/3649#issuecomment-5050598840 

## Are these changes tested?
Yes. Both unit tests and integration tests were writen

## Are there any user-facing changes?

Yes. `fast_forward_branch`.
See `mkdocs/docs/api.md` for more details. 

## Related Issues / Pull Requests
* Related to the closed PR https://github.com/apache/iceberg-python/pull/2434 (related issue https://github.com/apache/iceberg-python/issues/2433) ; my PR is only focused on implementing fast_forward
* Implemented one of the missing features mentioned in https://github.com/apache/iceberg-python/issues/737 from @sungwy 

## LLM Disclosure 
PR was written with the assistance of Claude Opus 4.7 (1M context) (effort: high)

It implemented the majority of the tests I defined. Specifically to assist with what fixtures to use and how best to write the unit tests in an atomic fashion appropriate for pyIceberg.

### CODE REVIEW

**Comment by rambleraptor** [comment]
I'm personally in favor of making those changes. I understand how a follow-up PR would be cleaner, but I worry about a situation where we merge this in + don't get around to the follow-up. 

I'm happy to do a review as-is. Which works better for you?

**Comment by JeroenSchmidt** [comment]
> I'm personally in favor of making those changes. I understand how a follow-up PR would be cleaner, but I worry about a situation where we merge this in + don't get around to the follow-up.
> 
> I'm happy to do a review as-is. Which works better for you?

Hi @rambleraptor 
Thank you for the response. Would it be possible for you to review it as is while I look into the open issue with chaining? Depending on the size of the change I can then add it as a commit or raise another PR while this one gets reviewed. 

I'm still getting familiar with the inner workings of this library so some guidance would help. 

Do you know if it's safe for me to construct the refs to point to in fast_forward by inspecting `self._transaction.table_metadata.refs` before any commits occur? In other words, are there any fail states or edge cases that I would need to consider if I start referencing refs mentioned in `self._transaction.table_metadata.refs` that are not committed yet?

**Comment by rambleraptor** [comment]
There's almost certainly a lot of edge cases, but we'll probably have to discover them as we go. I'll do a review and we can go from there.

**Comment by JeroenSchmidt** [comment]
Sounds good. I'll try come up with something and then see where it leads us. 😄 
I think I'll have something to share EoW / early next week depending on my other commitments and your review comments.

**Comment by JeroenSchmidt** [comment]
Hi @rambleraptor, 
I think I found a working solution that looks correct. What do you think?
See https://github.com/apache/iceberg-python/pull/3649/commits/57ebb876fc6e82399c45341bdbb006cc9b9b138f


I also added an [explicit integration test](https://github.com/apache/iceberg-python/pull/3649/commits/29a4f762e0b828263c9d30d52784d471b2b8747f) to validate everything works as expected.

Let me know if this solution looks ok and Ill proceed to implement some explicit tests for the `_effective_refs`.

**Comment by JeroenSchmidt** [comment]
Hi @rambleraptor,
Will you be able to have another look?

File: tests/table/test_manage_snapshots.py [hunk 1]
```diff
@@ -177,3 +183,234 @@ def test_set_current_snapshot_chained_with_create_tag(table_v2: Table) -> None:
...
+
+
+def test_fast_forward_branch_advances_to_descendant(table_v2: Table) -> None:
+    parent_snapshot_id = 3051729675574597004
+    child_snapshot_id = 3055729675574597004
+
+    # Create a lagging branch at the parent snapshot, then reset the mock so
+    # the next commit's updates are the fast-forward alone.
+    table_v2.catalog = MagicMock()
+    table_v2.catalog.commit_table.return_value = _mock_commit_response(table_v2)
```
Comment (rambleraptor) [inline tests/table/test_manage_snapshots.py:195]: You should inject in the ref directly here like you do in other tests (preserve_retention_fields)

```
table_v2.metadata.refs["lagging"] = SnapshotRef(
         snapshot_id=parent_snapshot_id,
         snapshot_ref_type="branch",
)
```

**Comment by rambleraptor** [comment]
I did a quick pass on this and it broadly looks correct! I'll take another look over the docs in particular on Monday. I want to make sure that our documentation is succinct and tells the appropriate amount of detail.

File: mkdocs/docs/api.md [hunk 2]
```diff
@@ -1483,6 +1483,72 @@ Remove an existing branch:
 table.manage_snapshots().remove_branch("dev").commit()
 ```
 
+#### Fast-forwarding a branch
+
+Fast-forward `from_branch` to point at the snapshot referenced by
+`to_ref`. `to_ref` may be a branch or tag; `from_branch` must be a
+branch. If `from_branch` does not yet exist it is created pointing at
```
Comment (rambleraptor) [inline mkdocs/docs/api.md:1490]: Worth adding that an auto-created from_branch gets default retention properties.

File: tests/table/test_manage_snapshots.py [hunk 3]
```diff
@@ -177,3 +183,234 @@ def test_set_current_snapshot_chained_with_create_tag(table_v2: Table) -> None:
     # The main branch should point to the same snapshot as the tag
     main_update = next(u for u in set_ref_updates if u.ref_name == "main")
     assert main_update.snapshot_id == snapshot_one
+
+
+def test_fast_forward_branch_advances_to_descendant(table_v2: Table) -> None:
```
Comment (rambleraptor) [inline tests/table/test_manage_snapshots.py:188]: Some test coverage gaps:

-  A successful fast-forward where to_ref is a tag
- A no-op

**Comment by Ferdinanddb** [comment]
Hi @JeroenSchmidt, since this looked stalled on a rebase I picked it up in #3900, keeping your API shape and intra-chain approach and addressing the open review comments there. Credited you in the PR description. Happy to close mine instead if you plan to continue this one.

**Comment by JeroenSchmidt** [comment]
Hi @Ferdinanddb 
Thanks for the message. I was coincidentally working on this on Friday. If you don't mind, I'd like to wrap it up still. I should have it pushed by Tuesday morning for a final review.

File: tests/table/test_manage_snapshots.py (hunk 3 above)
Comment (JeroenSchmidt) [reply tests/table/test_manage_snapshots.py]: Hi @rambleraptor 
I renamed the existing noop test and added 2 more to guard against some edge cases. 

**no-op tests**
* test_fast_forward_branch__when_refs_equal__does_not_commit (renamed)
* test_fast_forward_branch__when_noop_and_other_chained_updates__stages_only_the_other (new)
* test_fast_forward_branch__when_applied_twice_in_one_chain__second_is_noop (new)

For the tag case, I added a new test `test_fast_forward_branch__with_existing_from_branch__advances_to_descendant_tag`

**Comment by JeroenSchmidt** [comment]
Hi @rambleraptor 
Thanks again for the review. I addressed you comments. See the latest 2 commits. 

* I cleaned up and refined the docs. I also added a mermaid diagram for WAP. I hope you like them.
* Added 2 new noop tests and renamed the existing noop test. 
* Added a fast-forward test where to_ref is a tag
* Updated the test names to be more readable `test__{method}__with_{condition}__{outcome}`

File: mkdocs/docs/api.md [hunk 4]
```diff
@@ -1483,6 +1483,110 @@ Remove an existing branch:
...
+
+The use of branching & fast-forwarding enable the usage of the write-audit-publish (WAP) process:
+
+1. Writes proceed on a side branch
+2. Audit Validation runs against that branch
+3. Publish the new data by fast-forwarding the main branch
+
+```mermaid
+---
+title: Conceptually Illustration of the WAP Process
```
Comment (Fokko) [inline mkdocs/docs/api.md:1520]: Nice!

Image

File: pyiceberg/table/update/snapshot.py [hunk 5]
```diff
@@ -1232,6 +1267,61 @@ def _current_ancestors(self) -> set[int]:
...
+        Raises:
+            NoSuchSnapshotRefError: ``to_ref`` does not exist.
+            SnapshotRefTypeError: ``from_branch`` exists but is a tag.
+            NotAncestorError: ``from_branch``'s snapshot is not an ancestor of ``to_ref``'s snapshot.
+        """
+        refs = self._effective_refs()
+
+        if to_ref not in refs:
+            raise NoSuchSnapshotRefError(f"Ref does not exist: {to_ref}")
+        to_snapshot_id = refs[to_ref].snapshot_id
```
Comment (Fokko) [inline pyiceberg/table/update/snapshot.py]: nit, if you want to do a single lookup:
```suggestion
        if (to_snapshot_id := refs.get(to_ref)) is None:
            raise NoSuchSnapshotRefError(f"Ref does not exist: {to_ref}")
```
Another example below

File: pyiceberg/exceptions.py [hunk 6]
```diff
@@ -146,3 +146,15 @@ class WaitingForLockException(Exception):
...
+class NoSuchSnapshotRefError(ValueError):
+    """Raised when a named snapshot ref (branch or tag) does not exist."""
+
+
+class SnapshotRefTypeError(ValueError):
+    """Raised when an operation expects a branch and gets a tag (or vice versa)."""
+
+
+class NotAncestorError(ValueError):
+    """Raised when an operation requires ancestry between two snapshots and it does not hold."""
```
Comment (Fokko) [inline pyiceberg/exceptions.py:160]: At most places in the code, we default to `ValueError`, but I can see the value of this.

**Review by Fokko — APPROVED** [review]

**Comment by Fokko** [comment]
Thanks @JeroenSchmidt for working on this, it looks great!

File: pyiceberg/exceptions.py (hunk 6 above)
Comment (JeroenSchmidt) [reply pyiceberg/exceptions.py:160]: I'm glad you agree with this 😄 

I find that these kind of exception types help a lot when you need tighter flow control when using iceberg within the IO interface / framework layer of an application

**Review by rambleraptor — APPROVED** [review]
Yep, I think we're there. Thanks so much for the contribution!

**Comment by rambleraptor** [comment]
@JeroenSchmidt it looks like there's a single failing test (that's showing up on the Windows runners). My guess is that it's not Windows related (those issues usually involve file path issues). Can you take a look at that bug? Pushing up an empty commit will also re-kick the tests, so we can see the rest of them to confirm it's not Windows-only.

**Comment by JeroenSchmidt** [comment]
@rambleraptor 
I think it might be related to the NIT commit I added (without testing) https://github.com/apache/iceberg-python/pull/3649/commits/1a983736c48b773fc8349d5d2fe9496d50c8047b 

I'll see if I can find a fix. 

**Update:** Found the issue. Fix is in https://github.com/apache/iceberg-python/pull/3649/commits/fcf9faccdefc6672391e934848f52f5628959bae

**Comment by JeroenSchmidt** [comment]
@Fokko @rambleraptor It looks like the integration tests are failing unrelated to this PR. Is there something I need to do?

> target hive: failed to solve: process "/bin/sh -c apt-get update -qq &&     apt-get -qq -y install --no-install-recommends curl &&     apt-get clean &&     rm -rf /var/lib/apt/lists/*" did not complete successfully: exit code: 100

**Comment by Fokko** [comment]
Thanks @JeroenSchmidt for pinging, this should solve the issue: https://github.com/apache/iceberg-python/pull/3924 Can you pull in latest main?

**Comment by JeroenSchmidt** [comment]
Hi @Fokko thanks for the quick update. It looks like the PR you shared has been in a queued state for about an hour now. 

I'll have another look later today. Hopefully the merge queue is completed by then so that I can pull main into my branch.

**Comment by Fokko** [comment]
Thanks @JeroenSchmidt, not sure what's going on. It should merge within 20 minutes:

image

cc @kevinjqliu

**Comment by Ferdinanddb** [comment]
Hey @JeroenSchmidt can you retry to pull main on your branch when you have a minute? Sorry for the spam but I really want this feature to be available as I am waiting for it since a long time now.

**Comment by JeroenSchmidt** [comment]
Hi @Fokko @rambleraptor 
Thanks for resolving the blocker yesterday. I pushed the rebase. ~Hopefully it works now.~ looks like it works 😄 

FYI: @Ferdinanddb - 🤞 CI works now.

**Review by rambleraptor — APPROVED** [review]
I had approved this earlier. Thanks so much for doing this!

**Comment by rambleraptor** [comment]
@Fokko it looks like ASF runners are in a better place today. Mind putting this back into the merge queue?

**Comment by Fokko** [comment]
Sure thing! Let's go!
